src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +69,101 @@ def install_given_reqs(
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
+        # get a list of packages we can install in parallel
+        parallel_reqs_index = [
+            i for i, req in enumerate(requirements)
+            if not req.should_reinstall and req.is_wheel]
+
+        # install packages parallel
+        if parallel_reqs_index:
+            parallel_reqs = map_multiprocess_ordered(
+                partial(_single_install, install_args, suppress_exception=True),
+                [requirements[i] for i in parallel_reqs_index])
+        else:
+            parallel_reqs = []
+
+        # check the results from the parallel installation,
+        # and fill-in missing installations or raise exception
+        for i, req in enumerate(requirements):
+
+            # select the install result from the parallel installation
+            # or install serially now
+            if parallel_reqs_index and parallel_reqs_index[0] == i:
+                installed_req = parallel_reqs.pop(0)
+                parallel_reqs_index.pop(0)
             else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
+                installed_req = _single_install(
+                    install_args, req, suppress_exception=False)
+
+            # Now processes the installation result,
+            # throw exception or add into installed packages
+            if isinstance(installed_req, BaseException):
+                # Raise an exception if we caught one
+                # during the parallel installation
+                raise installed_req
+            elif isinstance(installed_req, InstallationResult):
+                installed.append(installed_req)
 
     return installed
+
+
+def _single_install(
+        install_args,                # type: _InstallArgs
+        requirement,                 # type: InstallRequirement
+        suppress_exception=False,    # type: bool
+):
+    # type: (...) -> Union[None, InstallationResult, BaseException]
+    """
+    Install a single requirement, returns InstallationResult
+    (to be called per requirement, either in parallel or serially).
+    Notice the two lists are of the same length
+    """
+
+    if requirement.should_reinstall:
+        logger.info('Attempting uninstall: %s', requirement.name)
+        with indent_log():
+            uninstalled_pathset = requirement.uninstall(
+                auto_confirm=True
+            )
+    else:
+        uninstalled_pathset = None
+
+    try:
+        requirement.install(
+            **install_args._asdict()
+        )
+    except (KeyboardInterrupt, SystemExit):
+        # always raise, we catch it in external loop
+        raise
```

Comment:
I think we can remove this block now since these are not `Exception`.

Reply:
If we do, then `ctrl-c` during the multi-process install will not be caught... wdyt?

Reply:
I’m not sure I undersntad. Isn’t catching them an re-raising immediately basically a no-op? Is there something I’m missing in a multiprocessing context?

Reply:
Ohh right, I forgot we reverted the `BaseException` to `Exception`... Yes you are correct now it is an no-op, I'll remove the redundant catch.